### PR TITLE
App/model links (add only)

### PR DESCRIPTION
**Pull request by nishadsingh1**
This PR addresses the third checkbox in #218. It only allows for adding links -- I'll work on a PR that allows for removing links after this gets edited and merged.

Given that users cannot yet remove links and that we limit apps to having at most 1 linked model, the only real difference this PR introduces to Clipper users is that app's candidate models (we now call them linked models) can be specified after app registration.

### CODE REVIEW

**Comment by nishadsingh1** [comment]
Looks like it's a formatting error: https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/463/consoleFull
"Found 1 C++ style violations.
Run format_code.sh to fix."

My code formatting script doesn't recommend any changes. Would one of you mind pushing up a formatting fix?

**Comment by nishadsingh1** [comment]
Thanks @dcrankshaw

**Comment by nishadsingh1** [comment]
"Build timed out (after 120 minutes). Marking the build as aborted.
Build was aborted" -- is this normal / should we just restart the build?

**Comment by nishadsingh1** [comment]
jenkins test this please

**Comment by nishadsingh1** [comment]
jenkins test this please

**Comment by nishadsingh1** [comment]
jenkins test this please

**Comment by nishadsingh1** [comment]
jenkins test this please

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
test this please

**Comment by nishadsingh1** [comment]
Somehow, tests passed this time without any functional changes (just added print statements):
```
[ RUN      ] RedisTest.SubscriptionDetectModelLinksAdd
[12:08:01.529][info]       [REDIS] Successfully issued command "FLUSHALL"
[12:08:01.529][info]       [REDIS] Successfully issued command "CONFIG SET notify-keyspace-events AKE"
[12:08:01.529][info]       [REDIS] Successfully issued command "SELECT 5"
[12:08:01.529][info]       [REDIS] Successfully issued command "HMSET my_app_name input_type doubles policy exp3_policy default_output 1.0 latency_slo_micros 10000"
[12:08:01.529][info]       [REDIS] SUBSCRIPTION STRING: __keyspace@7__:*
[12:08:02.030][info]   [REDISTEST] Before add_mode_links
[12:08:02.030][info]       [REDIS] Successfully issued command "SELECT 7"
[12:08:02.031][info]       [REDIS] Successfully issued command "SADD my_app_name model1"
[12:08:02.031][info]       [REDIS] MESSAGE: sadd
[12:08:02.031][info]   [REDISTEST] Received event
[12:08:02.031][info]   [REDISTEST] Confirmed event is SADD
[12:08:02.031][info]   [REDISTEST] Got lock in subscription callback
[12:08:02.031][info]   [REDISTEST] Before cv notify
[12:08:02.031][info]   [REDISTEST] After cv notify
[12:08:02.031][info]       [REDIS] MESSAGE: sadd
[12:08:02.031][info]   [REDISTEST] Received event
[12:08:02.031][info]   [REDISTEST] Confirmed event is SADD
[12:08:02.031][info]   [REDISTEST] Got lock in subscription callback
[12:08:02.031][info]   [REDISTEST] Before cv notify
[12:08:02.031][info]   [REDISTEST] After cv notify
[12:08:02.031][info]       [REDIS] Successfully issued command "SADD my_app_name model2"
[12:08:02.031][info]   [REDISTEST] After add_mode_links
[12:08:02.031][info]   [REDISTEST] Got lock outside of subscription callback
[12:08:02.031][info]   [REDISTEST] After waiting on cv
[       OK ] RedisTest.SubscriptionDetectModelLinksAdd (1527 ms)
```

Maybe I'll wait on getting this PR into its final format before worrying about why Jenkins tests are failing/passing on this code (since these occasional failures seem to be specific to this PR)

File: src/libclipper/src/redis.cpp
Comment (nishadsingh1) [inline src/libclipper/src/redis.cpp:461]: We're making multiple requests to Redis here instead of sending the whole set of `model_names` at once. This is because I couldn't find a way to add multiple elements to a set using the available `send_cmd_no_reply` infrastructure in one call. Let me know if there's a way to do this.

It's possible that no such way exists and sending a request for each new linked model is a bad practice. In this case, we could just have this REDIS_APP_MODEL_LINKS table store a mapping between app names and a string. The string would be a serialized list of model names (maybe with one of our predefined list delimiters). Then, whenever we want to fetch model links, we'd just deserialize the list. When adding new ones, we'd fetch the list, deserialize it, add to it, and re-serialize, and send it to Redis.

For now it's possible this isn't a big deal because we enforce that `model_names` has no more than 1 entry.

File: README.md
Comment (Corey-Zumar) [inline README.md]: This no longer conforms to the method prototype of `register_application`. We should remove `"feature_sum_model"`.

File: clipper_admin/clipper_manager.py
Comment (Corey-Zumar) [inline clipper_admin/clipper_manager.py]: How does "Allows the model with `model_name` to be used by the app with `app_name`" sound?

We should also note that `app_name` should be a preexisting application that has already been registered with clipper, while the model does not have to be registered.

File: clipper_admin/clipper_manager.py
Comment (Corey-Zumar) [inline clipper_admin/clipper_manager.py:396]: Remove extraneous print statement

File: src/management/src/management_frontend.hpp
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: Nit: `A model with name <model_name> is already linked to <app_name>` is cleaner.

File: clipper_admin/clipper_manager.py
Comment (Corey-Zumar) [inline clipper_admin/clipper_manager.py]: Replace "app" with "specified application".

File: examples/tutorial/tutorial_part_one.ipynb
Comment (Corey-Zumar) [inline examples/tutorial/tutorial_part_one.ipynb]: Nit: typo. `apo` >> `application`

File: examples/tutorial/tutorial_part_one.ipynb
Comment (Corey-Zumar) [inline examples/tutorial/tutorial_part_one.ipynb]: "In order to generate predictions, a Clipper application needs to be linked to a model. Let Clipper know that your..." is clearer.

File: examples/tutorial/tutorial_part_one.ipynb
Comment (Corey-Zumar) [inline examples/tutorial/tutorial_part_one.ipynb]: While we're here, let's remove the reference to the "feedback" endpoint, as we don't currently support it or incorporate it as part of the tutorial.

File: examples/tutorial/tutorial_part_one.ipynb
Comment (Corey-Zumar) [inline examples/tutorial/tutorial_part_one.ipynb]: Nit: `below code` >> `code below`

File: src/frontends/src/query_frontend.hpp
Comment (Corey-Zumar) [inline src/frontends/src/query_frontend.hpp]: Why do we reference `linked_model_names` as an`optional` here when `clipper::redis::get_linked_models` returns a vector?

File: src/frontends/src/query_frontend.hpp
Comment (Corey-Zumar) [inline src/frontends/src/query_frontend.hpp]: If `linked_model_names` does not need to be an optional, we can instead check whether or not it's empty.

File: src/libclipper/include/clipper/redis.hpp
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/redis.hpp]: Nit: Typo `am` >> `an`

File: clipper_admin/clipper_manager.py
Comment (Corey-Zumar) [inline clipper_admin/clipper_manager.py:399]: This results in `ValueError: No JSON object could be decoded` in the case that `app_name` is not a registered application.

File: src/frontends/src/query_frontend.hpp
Comment (Corey-Zumar) [reply src/frontends/src/query_frontend.hpp]: This method may not be necessary (see my comment below about whether or not it's a good idea to store link information locally)

File: src/frontends/src/query_frontend.hpp
Comment (Corey-Zumar) [inline src/frontends/src/query_frontend.hpp:278]: What's the rationale behind storing a copy of `linked_models_for_apps_` rather than querying Redis when a request is made for the data? These kinds of queries are probably infrequent and won't present an excessive load on the database. Unless there's a good reason, we should rely on Redis as a source of truth.

File: src/libclipper/src/json_util.cpp
Comment (Corey-Zumar) [inline src/libclipper/src/json_util.cpp]: This should be `for (auto const &str: str_array) {`. The reason being that using `auto` alone iterates by value, which I'm pretty sure incurs a copy.

File: src/libclipper/include/clipper/redis.hpp
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/redis.hpp]: Nit: `appname` >> `app_name`

File: src/libclipper/src/redis.cpp
Comment (Corey-Zumar) [inline src/libclipper/src/redis.cpp]: We should check that `result` is present before attempting to dereference it from the `optional` container.

File: src/libclipper/src/redis.cpp
Comment (Corey-Zumar) [reply src/libclipper/src/redis.cpp:461]: @dcrankshaw Any advice here?

It's true that this is fine for now because of the restriction that an application may only have one backing model. However, it would be nice to write this efficiently for future use cases when the restriction is loosened.

File: src/management/src/management_frontend.hpp
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: Nit: redundant `:=`

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
Functionality is almost complete! I really like the new link interface. Most of my comments are cosmetic.

**Comment by nishadsingh1** [comment]
- [x] Make sure models exist before linking them

**Comment by nishadsingh1** [comment]
Thanks @Corey-Zumar for the review -- the latest commit should address your comments and enforce that models are registered with Clipper before linking them to any apps.

Tests are passing locally for me; they're failing on Jeknins because the spark scala container isn't building correctly (fix attempt at #248), though of course there may be legitimate failures too that will show up later.

Even without the passing Jenkins tests, I think this is still ready for another pass-over.

**Comment by nishadsingh1** [comment]
Looks like it's hanging on `RedisTest. SubscriptionDetectModelLinksAdd`, even when print statements are added back in.

**Comment by shaneknapp** [comment]
i just checked the worker that this test ran on and it's got a TON zombie
processes.  this could be causing issues.

i'll need to reboot this machine to get rid of them.

File: clipper_admin/clipper_manager.py
Comment (Corey-Zumar) [inline clipper_admin/clipper_manager.py:395]: On second thought, let's re-add the call to `print(r.text)`. It's very helpful for user understanding - sorry!

File: src/libclipper/src/redis.cpp
Comment (Corey-Zumar) [inline src/libclipper/src/redis.cpp:457]: Currently, this method allows users to link an application of input type A to a model of input type B. We should verify that the input types are the same and fail if they aren't. Let's add a test for this as well.

File: src/libclipper/src/redis.cpp
Comment (Corey-Zumar) [reply src/libclipper/src/redis.cpp:461]: Resolved. This function is not on any critical paths, so making a few extra database queries will not have any significant impact on performance.

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
Almost there - just missing some input type verification during the linking process

File: src/libclipper/src/redis.cpp
Comment (nishadsingh1) [reply src/libclipper/src/redis.cpp:457]: Great catch :) I think it's correct to do this in `redis.cpp`, but we could also do it in the management front end. Just want to confirm that `redis.cpp` is the right place for it - thoughts?

File: clipper_admin/clipper_manager.py
Comment (nishadsingh1) [reply clipper_admin/clipper_manager.py:395]: Np at all

File: src/libclipper/src/redis.cpp
Comment (nishadsingh1) [reply src/libclipper/src/redis.cpp:457]: Actually, I'm going to redistribute the validation responsibilities.
Management frontend:
- Confirm that the user supplied only one model_name
- Make sure that there will only be one link

Redis:
- Confirm that an app with specified app name exists
- Confirm that a model exists for each specified model name
- Confirm input types match

It seems like we'd want the checks that guarantee important logic to exist at the single entry point to the database (redis.cpp). Let me know if that sounds reasonable.

--------------------
Edit:  It looks like we aren't propagating errors up from any of the functions in `redis.cpp`, so there isn't a standard way to handle them (and access their specific error messages) from the management frontend function.

We should definitely still let the user know when their specified app/model doesn't exist or if the input types don't match, instead of returning some nondescriptive error message in all cases. In order to give the appropriate error message, we need the management frontend function to make the checks that are under the 'Redis' section above.

Should we just duplicate the checks like so?:

Management frontend:
- Confirm that the user supplied only one model_name
- Make sure that there will only be one link
- Confirm that an app with specified app name exists
- Confirm that a model exists for each specified model name
- Confirm input types match

Redis:
- Confirm that an app with specified app name exists
- Confirm that a model exists for each specified model name
- Confirm input types match

I think this is probably okay since adding model links isn't on a critical path.

--------------------
Edit 2: We don't seem to do much validation at all in `redis.cpp`. Is this intentional / does this mean that we should only do this validation in the management frontend?

**Comment by nishadsingh1** [comment]
**Problematic scenario 1:**
- Register a model with name 'M', version '1', and input_type 'ints'.
- Register an app with name 'A' and input_type 'ints'
- Add a model link between 'M' and 'A'. Input types match, both apps exist, only one model name was specified and the app had no previous links: this request is successful.
- Register a model with name 'M', version '2', and input_type 'doubles'.

Queries to 'A' route to the latest version of 'M'. In this case, that would send requests to 'M:2', which has an incompatible input_type. Under the hood, registering 'M:2' set the current version of 'M' to '2'. This was the source of the problem.

**Problematic scenario 2**
- Register a model with name 'M', version '1', and input_type 'ints'.
- Register a model with name 'M', version '2', and input_type 'doubles'.
- Register an app with name 'A' and input_type 'doubles'
- Add a model link between 'M' and 'A'. Input types match, both apps exist, only one model name was specified and the app had no previous links: this request is successful.
- Set the current model version to '1'. 

Queries to 'A' route to 'M:1' which has an incompatible input_type.

So, it seems like we need to make sure that every time we set the current model version of a model, we're not creating some conflict with an existing linked application with respect to input type.

**Where to do this validation?**
If we put it in management frontend, we'd need to do the check in both `add_model` and `set_model_version`.

If we put it in redis.cpp's, we'd only need to do the check in `set_current_model_version`.

I like the latter option. If every time we change a model's model version, shouldn't we just have the check be in the function (or on a guaranteed path to the function ) that makes the change? Otherwise we run the risk of forgetting to run these checks with each usage of redis' `set_current_model_version` and creating bugs.

------
Update: we'll have all checks be in Management Frontend for now and move them to Redis in another PR

File: src/libclipper/src/redis.cpp
Comment (nishadsingh1) [reply src/libclipper/src/redis.cpp:457]: We resolved to just have all the checks be in the Management Frontend and to move them to Redis in another PR

**Comment by nishadsingh1** [comment]
Updates on the latest change:
- Addressed Corey's comments
- Made sure that input types matched for new links
- Made sure that when adding a new model, there exist no previous versions were linked to apps under different input types
- Made sure that when setting a new model version, there exist no links to applications under a the current model's input type that would be incompatible under the version change
- In the process of the latter two points, rewrote the frontend set_model_version code to match the form of our other frontend endpoints (plus, the change was necessary to avoid some uglier code changes to accommodate for the above checks)

File: src/management/src/management_frontend.hpp
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: Remove comment. Are there outstanding changes that need to be made?

File: src/management/src/management_frontend.hpp
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: The `check_updated_model_consistent_with_app_links` method is being invoked with `new_model_version` as the second parameter, as opposed to an `input_type` value.

Also, `attempt_model_version_update` calls the same method again. Does it need to be called in both places?

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
one more issue & doc cleanup. ready to merge when these changes are made and tests pass.

**Review by Corey-Zumar — APPROVED** [review]
Looks good. Will merge when tests pass. Looks like there's an issue with pyspark model deployment.

**Comment by dcrankshaw** [comment]
Actually hold off on merging this. This is not backwards compatible with the 0.1 release and will cause problems due to #244

**Comment by dcrankshaw** [comment]
Actually you can go ahead and merge it. I turned off the automated builds on docker hub until we can address #244. I'll prioritize that and push out a new 0.1.x release that manages docker versions properly.

**Pull request merged**
